trip_app/trips/views.py

```diff
@@ -69,21 +80,29 @@ def get_queryset(self):
 
     @action(detail=True, methods=["post"])
     def reserve(self, request, pk=None):
-        """ Reserve a trip by sending a POST to api/trip/<trip_id>/reserve/."""
+        """ Reserve a trip via POST api/trips/<trip_id>/reserve/. """
         trip = self.get_object()
         user = request.user
         if user == trip.driver:
             return Response(
                 data="Driver can't be a passenger at the same time",
                 status=status.HTTP_400_BAD_REQUEST,
             )
+        if user.requests.filter(trip_id=trip.id).exists():
+            return Response(
+                data="You have already requested this trip",
+                status=status.HTTP_400_BAD_REQUEST,
+            )
         if trip.free_seats:
```

Comment:
1. I'd think of this as validation, which should go ahead of successfull scenario. In that way you break the flow if something is not valid, and you move on otherwise. You also earn the possibility to nest code less with such a layout.
1. seems like this `if` (and the next one) have to do with the business-logic. If so, the view isn't the right place for that. @Jamim might advise here.

Reply:
I moved this logic into Trip model layer, I believe code is still not perfect but this should serve as a start point for further refactoring... I'd appreciate any advice on improving the code...